File: shared/types/capabilities.ts
import type { SpeedResult } from './speed'

export const CAPABILITY_LEVELS = [
  'excellent',
  'good',
  'fair',
  'poor',
  'unsupported',
] as const

export type CapabilityLevel = (typeof CAPABILITY_LEVELS)[number]

export interface CapabilityRequirement {
  readonly minDownloadMbps: number
  readonly minUploadMbps: number
  readonly maxLatencyMs: number
  readonly maxJitterMs: number
}

export const CAPABILITY_DEFINITIONS = [
  {
    id: 'browsing',
    label: 'Web Browsing',
    description: 'General web, email, social media',
    icon: '🌐',
    requirements: { minDownloadMbps: 5, minUploadMbps: 1, maxLatencyMs: 100, maxJitterMs: 30 },
  },
  {
    id: 'streaming-hd',
    label: 'HD Streaming',
    description: '1080p video streaming',
    icon: '📺',
    requirements: { minDownloadMbps: 10, minUploadMbps: 3, maxLatencyMs: 100, maxJitterMs: 30 },
  },
  {
    id: 'streaming-4k',
    label: '4K Streaming',
    description: 'Ultra HD video streaming',
    icon: '🎬',
    requirements: { minDownloadMbps: 25, minUploadMbps: 5, maxLatencyMs: 80, maxJitterMs: 20 },
  },
  {
    id: 'video-calls',
    label: 'Video Calls',
    description: 'Zoom, Teams, Google Meet',
    icon: '📹',
    requirements: { minDownloadMbps: 8, minUploadMbps: 3, maxLatencyMs: 150, maxJitterMs: 40 },
  },
  {
    id: 'gaming',
    label: 'Online Gaming',
    description: 'Low-latency multiplayer games',
    icon: '🎮',
    requirements: { minDownloadMbps: 15, minUploadMbps: 5, maxLatencyMs: 50, maxJitterMs: 15 },
  },
  {
    id: 'cloud-gaming',
    label: 'Cloud Gaming',
    description: 'GeForce NOW, Xbox Cloud',
    icon: '☁️',
    requirements: { minDownloadMbps: 35, minUploadMbps: 10, maxLatencyMs: 40, maxJitterMs: 10 },
  },
  {
    id: 'large-downloads',
    label: 'Large Downloads',
    description: 'Game updates, file transfers',
    icon: '⬇️',
    requirements: { minDownloadMbps: 50, minUploadMbps: 10, maxLatencyMs: 200, maxJitterMs: 50 },
  },
  {
    id: 'live-streaming',
    label: 'Live Streaming',
    description: 'Twitch, YouTube live broadcasting',
    icon: '📡',
    requirements: { minDownloadMbps: 10, minUploadMbps: 10, maxLatencyMs: 100, maxJitterMs: 30 },
  },
] as const

export type CapabilityId = (typeof CAPABILITY_DEFINITIONS)[number]['id']

export interface CapabilityDefinition {
  readonly id: CapabilityId
  readonly label: string
  readonly description: string
  readonly icon: string
  readonly requirements: CapabilityRequirement
}

export interface CapabilityRating {
  readonly id: CapabilityId
  readonly label: string
  readonly description: string
  readonly icon: string
  readonly level: CapabilityLevel
}

export interface SpeedTestResult {
  readonly internet: SpeedResult
  readonly capabilities: readonly CapabilityRating[]
  readonly linkCapacityMbps: number | null
}

export const CAPABILITY_LEVEL_LABELS: Readonly<Record<CapabilityLevel, string>> = {
  excellent: 'Excellent',
  good: 'Good',
  fair: 'Fair',
  poor: 'Poor',
  unsupported: 'Unsupported',
}

File: crates/rove-core/src/capabilities.rs
use crate::types::{CapabilityRating, SpeedResult};

struct Def {
    id: &'static str,
    label: &'static str,
    description: &'static str,
    icon: &'static str,
    min_down: f64,
    min_up: f64,
    max_latency: f64,
    max_jitter: f64,
}

const DEFS: [Def; 8] = [
    Def { id: "browsing", label: "Web Browsing", description: "General web, email, social media", icon: "🌐", min_down: 5.0, min_up: 1.0, max_latency: 100.0, max_jitter: 30.0 },
    Def { id: "streaming-hd", label: "HD Streaming", description: "1080p video streaming", icon: "📺", min_down: 5.0, min_up: 1.0, max_latency: 100.0, max_jitter: 30.0 },
    Def { id: "streaming-4k", label: "4K Streaming", description: "Ultra HD video streaming", icon: "🎬", min_down: 25.0, min_up: 2.0, max_latency: 80.0, max_jitter: 20.0 },
    Def { id: "video-calls", label: "Video Calls", description: "Zoom, Teams, Google Meet", icon: "📹", min_down: 5.0, min_up: 4.0, max_latency: 150.0, max_jitter: 30.0 },
    Def { id: "gaming", label: "Online Gaming", description: "Low-latency multiplayer games", icon: "🎮", min_down: 5.0, min_up: 2.0, max_latency: 50.0, max_jitter: 15.0 },
    Def { id: "cloud-gaming", label: "Cloud Gaming", description: "GeForce NOW, Xbox Cloud", icon: "☁️", min_down: 25.0, min_up: 3.0, max_latency: 40.0, max_jitter: 10.0 },
    Def { id: "large-downloads", label: "Large Downloads", description: "Game updates, file transfers", icon: "⬇️", min_down: 50.0, min_up: 2.0, max_latency: 200.0, max_jitter: 50.0 },
    Def { id: "live-streaming", label: "Live Streaming", description: "Twitch, YouTube live broadcasting", icon: "📡", min_down: 5.0, min_up: 6.0, max_latency: 100.0, max_jitter: 30.0 },
];

fn rate(speed: &SpeedResult, def: &Def) -> &'static str {
    if speed.download_mbps < def.min_down * 0.5 || speed.latency_ms > def.max_latency * 2.0 {
        return "unsupported";
    }
    let meets = speed.download_mbps >= def.min_down
        && speed.upload_mbps >= def.min_up
        && speed.latency_ms <= def.max_latency
        && speed.jitter_ms <= def.max_jitter;
    if meets {
        let excellent =
            speed.download_mbps >= def.min_down * 2.0 && speed.latency_ms <= def.max_latency * 0.5;
        return if excellent { "excellent" } else { "good" };
    }
    if speed.download_mbps >= def.min_down * 0.7 {
        "fair"
    } else {
        "poor"
    }
}

pub fn assess(speed: &SpeedResult) -> Vec<CapabilityRating> {
    DEFS.iter()
        .map(|def| CapabilityRating {
            id: def.id.into(),
            label: def.label.into(),
            description: def.description.into(),
            icon: def.icon.into(),
            level: rate(speed, def).into(),
        })
        .collect()
}
